tests/test_config_model.py

```diff
@@ -11,6 +11,40 @@ def test_pydantic_model(self):
             CONFIG_MODEL.zones["US-NW-PSCO"].parsers.get_function("production")
         )
 
+    # Add well-known sources that don't require config-based references here
+    GLOBAL_SOURCE_REFERENCES = {
+        "BEIS 2021",
+        "CEA 2022",
+        "EIA 2020/BEIS 2021",
+        "EU-ETS 2021",
+        "EU-ETS, ENTSO-E 2021",
+        "IEA 2019",
+        "IEA 2020",
+        'Oberschelp, Christopher, et al. "Global emission hotspots of coal power generation."',
+    }
+
+    def test_zone_sources(self):
+        for measurement_basis, model in CO2EQ_CONFIG_MODEL:
+            for zone_key, zone_modes in model.emission_factors.zone_overrides.items():
+                for mode, estimate in zone_modes or ():
+                    if estimate is None:
+                        continue
+                    estimates = estimate if isinstance(estimate, list) else [estimate]
+                    for estimate in estimates:
+                        self.assertIsNotNone(
+                            estimate.source,
+                            msg=f"Source is required for {mode} in {zone_key}",
+                        )
+                        for source in estimate.source.split(";"):
+                            source = source.strip()
+                            if source.startswith("assumes"):
+                                continue
+                            if source.startswith("Electricity Maps"):
+                                continue
+                            if source in self.GLOBAL_SOURCE_REFERENCES:
+                                continue
+                            self.assertIn(source, CONFIG_MODEL.zones[zone_key].sources)
```

Comment:
This line here has a type error and prints the following using pylance:
> Argument of type "dict[str, Source] | None" cannot be assigned to parameter "container" of type "Iterable[Any] | Container[Any]" in function "assertIn"

Reply:
Ok; and it seems that this is due to the fact that the field is optional (an `x in foo` check is invalid if `foo` is `None`).  But it _is_ correct for it to be an optional field.  And `pyright` isn't convinced-enough by adding a `assertIsNotNone` on the previous line to eliminate the `| None` possibility.

Thinking about how to resolve this.. (adding `or {}` to the end of this line would be a quick although not-ideal way).

Reply:
> And pyright isn't convinced-enough by adding a assertIsNotNone on the previous line to eliminate the | None possibility.

See microsoft/pyright#2007 for the (reasonable) rationale behind that.  However, the workaround suggested (add an `assert ... is not None` - attempted locally as `assert CONFIG_MODEL.zones[zone_key].sources is not None`) does not seem to be working for me.  I'll try to figure out whether that's a `pyright` bug.

Reply:
Yep, I think there could be a `pyright` bug here.  `assert .. is not None` doesn't seem to preserve the asserted non-empty condition in the case where attribute access is used here:

```py
                            zone_sources = CONFIG_MODEL.zones[zone_key].sources
                            assert CONFIG_MODEL.zones[zone_key].sources is not None
                            assert zone_sources is not None
                            self.assertIn(source, CONFIG_MODEL.zones[zone_key].sources)  # pyright complains here
                            self.assertIn(source, zone_sources)  # pyright does not complain here
```

Reply:
Hmm.  Maybe that's as-expected - possibly because the `pydantic.BaseModel` that `Zone` derives from implements `__getattr__`, so the results of attribute access could change from call to call.

Reply:
Ok, this should now be resolved (96974de4253351bdfdd14af1f8d68c9a9c559316 - hold the container's value in a local variable and use an `assert ... is not None` to help `pyright` narrow the typechecking for it).